Amplify/Categories/DataStore/Model/PropertyPath.swift

```diff
@@ -0,0 +1,126 @@
+/// structure of a model with static typing, so developers can reference model
+/// properties in queries and other functionality that require them.
+public protocol PropertyPath {
+
+    /// Access the property metadata.
+    ///
+    /// **Implementation note:** this function is in place over an implicit accessor over
+    /// a property named `metadata` in order to avoid name conflict with the actual property
+    /// names that will get generate from the `Model`.
+    ///
+    /// - Returns the property metadata, that contains the name and a reference to its parent.
+    func getMetadata() -> PropertyPathMetadata
+}
+
+/// This is a specialized protocol to indicate the property is a container of other properties,
+/// i.e. a `struct` representing another `Model`.
+///
+/// - SeeAlso: `ModelPath`
+public protocol PropertyContainerPath : PropertyPath {
+
+    /// 
+    func getKeyPath() -> String
+
+    /// Must return a reference to the type containing the properties
+    func getModelType() -> Model.Type
+
+}
+
+extension PropertyContainerPath {
+    
+    public func getKeyPath() -> String {
+        var metadata = getMetadata()
+        var path = [String]()
+        while let parent = metadata.parent {
+            path.insert(metadata.name, at: 0)
+            metadata = parent.getMetadata()
+        }
+        return path.joined(separator: ".")
+    }
+}
+
+/// Represents a scalar (i.e. data type) of a model property.
+public struct FieldPath<ValueType>: PropertyPath {
+    private let metadata: PropertyPathMetadata
+
+    init(name: String, parent: PropertyPath? = nil) {
+        self.metadata = PropertyPathMetadata(name: name, isCollection: false, parent: parent)
+    }
+
+    public func getMetadata() -> PropertyPathMetadata {
+        return metadata
+    }
+}
+
+/// Represents the `Model` structure itself, a container of property references.
+///
+/// - Example:
+/// ```swift
+/// class PostModelPath : ModelPath<Post> {}
+///
+/// extension ModelPath where ModelType == Post {
+///   var id: FieldPath<String> { id() }
+///   var title: FieldPath<String> { string("title") }
+///   var blog: ModelPath<Blog> { BlogModelPath(name: "blog", parent: self) }
+/// }
+/// ```
+open class ModelPath<ModelType: Model> : PropertyContainerPath {
+
+    private let metadata: PropertyPathMetadata
+
+    public init(name: String = "root", isCollection: Bool = false, parent: PropertyPath? = nil) {
+        self.metadata = PropertyPathMetadata(name: name, isCollection: isCollection, parent: parent)
+    }
+
+    public func getMetadata() -> PropertyPathMetadata {
+        return metadata
+    }
+
+    public func getModelType() -> Model.Type {
+        return ModelType.self
+    }
+
+    public func isRoot() -> Bool {
+        return metadata.parent == nil
+    }
+
+    public func id(_ name: String = "id") -> FieldPath<String> {
+        FieldPath(name: name, parent: self)
+    }
```

Comment:
Is there a special use case for this or just visually appealing to have it codegenerated as "id"? When hand-writing the schema with composite keys, both keys were not "id" so they are both `string(<fieldName>)`, which means in this case that the primary key of a model with composite key doesn't show which of the fields are its primary keys in the ModelPath extension. Should we just remove this completely? Currently, I haven't added the check for "is primaryKey and field value is 'id'" and simply generated "string(<fieldName>)" for all string fields including properties called "id"

Reply:
That's a good callout. It made sense before when we had ids based on the naming convention. That not being the case anymore calls for a different approach, I think we can simply use the `string()` function.